### ISSUE TITLE
Bucketing strategy not working properly for multiple GPUs? Cyclical training loss

### ISSUE THREAD

**Issue by whrichd** [NVIDIA/NeMo#6252]
**Describe the bug**

It's a similar observation to #6084, but since that one is closed and it's slightly different (plus I have some additional information) I'd like to ask again if this is a bug.

I looked at my training loss / batch WER when using bucketing, and found that the loss and WER follow the same pattern for every epoch, even though I specified `fully_randomized` as the bucketing strategy. Later I discovered that because used `pl.seed_everything()`, all the random generators were initialized the same and therefore all the GPUs use the same buckets. I changed this (not seeding now), and printed the `shuffled order` and `dataset_idx` in `RandomizedChainDataset`'s iter method. I can see that now there's some randomization happening when the datasets are loaded, but the train_loss graph still looks exactly the same as before. If I track the loss in the terminal progress bar, I can also see that for each epoch the loss is about in the same range at a certain point, e.g. 25% into the epoch, the loss is always around 100 whereas 75% it's about 50.

My questions are:
- is this desired behavior? I would expect that for example for 2 buckets, instead of following a pattern every epoch, it would have cycle through a pattern every half epoch if it's fully randomized. And even if it's synced randomized, it should be different every epoch.
- does it have to do with the `global_rank` and how the datasets are loaded at each GPU? I'm not familiar with this and it would be good if someone can explain how this works. Because now I see that when I print the `shuffled_order` at the start of every epoch, and I have 8 GPUs, it is printed 8x8 = 64 times. Is this `num_gpus` x `num_workers`?
- I would have expected the randomization of bucketing to still work despite using a random seed for my experiments. Is that doable?

Some graphs:

For training with 2 buckets, data size about 1000 hrs.
Training loss of the same training, orange is when I used random seed, blue is when I stopped the first run, removed random seed and continued from checkpoint. I see random dataset orders printed out, but the the train loss is exactly the same patter every epoch despite the randomization.
<img width="827" alt="image" src="https://user-images.githubusercontent.com/43883260/226186308-3826aee6-e91c-45bf-bf55-6c7d91d651f5.png">

Training batch WER for the same run. It looks slightly `smoothed out` after I removed the random seed, but I would have expected 2 peaks per epoch for example.
<img width="709" alt="image" src="https://user-images.githubusercontent.com/43883260/226186561-56244d11-5b06-488e-8c98-662d36ecd59f.png">

For training with 4 buckets
<img width="428" alt="image" src="https://user-images.githubusercontent.com/43883260/226186732-45ac67fe-ccd6-4faa-b95e-ee1171cc6503.png">
Here I started the training run again without the random seed (blue line) and with a higher batch size. But still it seems like I can clearly see where one bucket ends and another bucket starts because they have different lengths and therefore different average loss. Even though in the terminal during one epoch I see different `dataset_idx` printed out at different times. This training also doesn't converge, I'm wondering if it only has to do with the longer audio or because of the lack of randomization as well.

**Steps/Code to reproduce bug**

used `[speech_to_text_ctc_bpe.py](https://github.com/NVIDIA/NeMo/blob/a0f584a4480165ff144601336280a2276b43d4bd/examples/asr/asr_ctc/speech_to_text_ctc_bpe.py)` to run the training,

Cfg is conformer_ctc_bpe_streaming.yaml.
4 buckets: batch size = 24, 2 buckets batch size = 60 (bucketing batch size)
8 gpus
8 workers
lr: 2
warmup 10000 / 20000

**Expected behavior**

- Training loss should not be uniformly cyclical per epoch training using multiple GPUs and fully randomized bucketing.

**Environment overview (please complete the following information)**

 - Environment location: GCP
 - Method of NeMo install: followed the instruction in the readme to install in a conda env. First using pip install, then pip install -e for printing out the dataset_idx.
 - 8 GPUs

**Environment details**

python: 3.8.0
pytorch: 1.24.1
pytorch-lightning: 1.8.6
nemo-toolkit: 1.15.0

**Additional context**

GPU: A100

**Comment by VahidooX** [comment]
Thanks for reporting the bug!
I created a PR to fix the cases where pl.seed_everything() is being used here:
https://github.com/NVIDIA/NeMo/pull/6254

We used to observe this behavior sometimes in the past in our trainings but as the test and validation metrics were fine, we just ignored it. Not sure why it happens, maybe something is wrong with how we gather metrics over GPUs. I will try to look into it and get back to you. In the meantime, I think you can continue your training as it is unlikely to affect your results. How are your validation metrics like WER on test/validation?

>>>I see that when I print the shuffled_order at the start of every epoch, and I have 8 GPUs, it is printed 8x8 = 64 times.
Would you please show me exactly where you put this print?
It should be just 8 times in the output if you have printed that in the dataset creation.

**Comment by whrichd** [comment]
I see! Yeah my next thought is also that it might be due to metrics reporting.

> How are your validation metrics like WER on test/validation?

Well the reason I looked into this is because initially the model is not converging at all (small conformer streaming), but it looks like it is working now that I switched to shorter utterances (the two bucket case) and the validation WER is dropping normally (although val WER is much lower than train WER).

This is how I print the info:
which results in 64 lines of the shuffled order.
And random multiples of 8 for the dataset_idx depending on the order.
```
    def __iter__(self):
        shuffled_order = self.rnd_gen.permutation(len(self.datasets))
        print(f"bucketing shuffled_order is : {shuffled_order}")
        for dataset_idx in shuffled_order:
            print(f"bucketing dataset: {dataset_idx}")
            d = self.datasets[dataset_idx]
            assert isinstance(d, IterableDataset), "ChainDataset only supports IterableDataset"
            for x in d:
                yield x
```

**Comment by titu1994** [comment]
What are the bucket durations ? For example, what is the max duration in the final bucket ? If the variance between buckets is large, you will see significant deviation between loss of each bucket (unless you use "mean_volume" loss normalization instead of mean_batch. 

Also, something you should if your durations are too different is progressively learning in stages. First stage use just 2 buckets, next 4 and next all buckets each time using the model learned from the last training stage as init (stage 0 can be a pretrained conformer CTC small from pretrained checkpoints or can be random init if following academic protocol)

**Comment by whrichd** [comment]
Thanks for the tip! The max duration is 30 secs in the last bucket, but actually from what I'm seeing now, it's not just the duration because with just my in domain data, it doesn't converge even with just the shorter duration subset (<15s). But with Fisher (similar domain) added it starts to converge...I can't figure out what else in the data can cause such a thing. 

> stage 0 can be a pretrained conformer CTC small from pretrained checkpoints or can be random init if following academic protocol

May I ask what academic protocol you are talking about? I have fine-tuned models with your pretrained checkpoints and this is the first time to train from scratch because of specific model architecture needs, i.e. streaming model has different convolution layers which prevented pretrained weights being loaded. Unless there's some way to only load some layers not others? And I find training from random weights to be really difficult as explained above. Learns public datasets fine, but not if using in domain data alone.

**Comment by VahidooX** [comment]
Can you try model.ctc_reduction=mean_volume? It would normalize the loss based on length and would give you a smoother loss.
If WER did not do down from 100%, you may increase warmup from 10k to 20k then.
